src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+// @flow
+
+import { traverseAst } from "./utils/ast";
+import { getSource } from "./sources";
+
+import type { Source, SourceId, Location, SourceScope } from "debugger-html";
+import type { NodePath, Node, Location as BabelLocation } from "babel-traverse";
+
+export type ScopeDeclaration = {
+  type: string,
+  start: Location,
+  end: Location,
+  bindings: any
+};
+
+export type ScopeDeclarations = ScopeDeclaration[];
+
+type GeneratedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: any,
+  children: ?(GeneratedScope[])
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  scopes: Array<TempScope>,
+  loc: BabelLocation,
+  names: any
+};
+
+function _createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  let result = {
+    type,
+    parent,
+    scopes: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.scopes.push(result);
```

Comment:
we are making a circular reference here. that makes me a bit nervous -- what is the goal here?